specifications/unified-provider.org

```diff
@@ -0,0 +1,84 @@
+| is_provider_ready() -> bool                | checks if provider is ready w/o installation or configuration             |                                                       |
+| setup_provider(echoer) -> None             | installs/configures provider, allowing for user prompting                 | ProviderInstallationError, ProviderConfigurationError |
+| execute_step(step: steps.Step) -> None     | executes (up to and including) step                                       | NotImplementedError                                   |
+| expose_prime(path: pathlib.Path) -> None   | mount project =prime= directory to host at specified path                 | NotImplementedError                                   |
+| snap() -> List[pathlib.Path]               | snap project (building as required), writing snap(s) to host project dir. |                                                       |
+| clean_parts(part_names: List[str]) -> None | clean project parts                                                       | NotImplementedError                                   |
+| clean() -> None                            | clean all project artifacts, including build-instance, if applicable.     |                                                       |
+| shell() -> None                            | launch interactive shell with build-instance, if applicable.              | NotImplementedError                                   |
+|--------------------------------------------+---------------------------------------------------------------------------+-------------------------------------------------------|
+
+*** Summary of changes:
+- Rename =ensure_provider()= to =is_provider_ready()=, scoping to see if
+  provider is ready without needing installation or further configuration.
+  Remove exceptions.  CLI will invoke setup_provider() if terminal is
+  interactive, allowing user to perform any install/configuration as necessary.
+- Remove =mount_project()=. Project path will be provided to __init__() and will
+  be verified/mounted as required.
+- Replace =pack_project()= with =snap()=, adding the implication that build steps
+  are to be executed as required (and packing may not be done by Snapcraft).
+- Rename =clean()= to =clean_parts()=, to improve clarity of scope.
+- Rename =clean_project()= to =clean()=, to improve clarity of scope.
+
+** Migrate current logic into new providers:
+- HostProvider
+- ManagedHostProvider, inherits much of HostProvider.
+- LaunchpadProvider
+- RemoteLXDProvider, inherits much of LXDProvider, overrides the (now internal)
+  equivalent of mount_project() behavior with a rsync from host to remote.
+
+* CLI Handling
+
+** Provider NotImplementedError
+
+Certain commands will not supported by all providers. E.g.
+- =snapcraft try= is not compatible with Launchpad, Remote LXD, Host
+- =snapcraft <step>= is not compatible with Lauchpad, Remote LXD
+- =snapcraft clean <parts>= is not compatible with Launchpad, Remote LXD
```

Comment:
these latter two would be supported with remote-lxd in the future, we will just progressively implement. Care to describe that so the full vision is presented?

There is also no comment about --shell, --shell-after and --debug

Reply:
Ah yes, good points.

I'll write it such that Remote LXD will support <step> and <parts>.  I'm actually not sure why I wrote it this way :thinking: 

Will add that Launchpad will not support --shell, --shell-after and --debug.